### ISSUE TITLE
Failed conversion of hpc coordinates using SkyCoord with frames.Helioprojective

### ISSUE THREAD

**Issue by WaaallEEE** [sunpy/sunpy#2947]
I don't know if that is an expected behaviour or a bug, but the documentation could not answer:

Trying to convert helioprojective to pixel coordinate using SkyCoord: If I instantiate `SkyCoord` by initializing with `obstime` and `frame=sunpy.coordinates.frames.Helioprojective`, I have NaN for values close to the limb, whereas initializing it using a frame=`sunpy.map.Map` object gives me the expected pixel coordinates.

This will reproduce the problem: 

```
import astropy.units as u
from astropy.coordinates import SkyCoord
from sunpy.map import Map
import sunpy.data.sample
from sunpy.coordinates import frames

aia_map = Map(sunpy.data.sample.AIA_171_IMAGE)

# Try to convert coordinates given in helioprojective cartesian coordinates to pixel coordinates.
# Regardless of the coordinates, as long as it's in the image, we should get finite values, regardless of
# being off limb: e.g: filament eruption, CME, etc... there are always stuff we look at beyond the limb.

# Take coordinates not far from the limb
xy_hpc = [-167*u.arcsec, 930*u.arcsec]

# Convert to pixel coordinates
xy_coord1 = SkyCoord(*xy_hpc, frame=aia_map.coordinate_frame)
print(aia_map.world_to_pixel(xy_coord1))
# PixelPair(x=<Quantity 439.71987887 pix>, y=<Quantity 897.80294635 pix>)
xy_coord2 = SkyCoord(*xy_hpc, obstime=aia_map.date, frame=frames.Helioprojective)
print(aia_map.world_to_pixel(xy_coord2))
# PixelPair(x=<Quantity 439.73259955 pix>, y=<Quantity 897.73173924 pix>)

# Let's move up by 1 arcsec on the y axis...
xy_hpc = [-167*u.arcsec, 931*u.arcsec]

# Convert to pixel coordinates
xy_coord1 = SkyCoord(*xy_hpc, frame=aia_map.coordinate_frame)
print(aia_map.world_to_pixel(xy_coord1))
# PixelPair(x=<Quantity 439.71887043 pix>, y=<Quantity 898.21913623 pix>)
xy_coord2 = SkyCoord(*xy_hpc, obstime=aia_map.date, frame=frames.Helioprojective)
print(aia_map.world_to_pixel(xy_coord2))
# /Users/rattie/anaconda/envs/sdo_spikes/lib/python3.6/site-packages/astropy/units/quantity.py:463: RuntimeWarning: invalid value encountered in sqrt
#   result = super().__array_ufunc__(function, method, *arrays, **kwargs)
# PixelPair(x=<Quantity nan pix>, y=<Quantity nan pix>)
```

What am I missing? What does the sunpy AIA Map provides more to the SkyCoord object that the `obstime` and frame=`sunpy.map.Map` does not, since in the end, the same sunpy Map object is used in both cases when using 'world_to_pixel()'? 
Is this expected behavior?

**Comment by wafels** [comment]
Maybe the AIA headers are more accurate, taking into account the position of the spacecraft. I think the default assumption assumes an observing position at the center of the Earth.

**Comment by Cadair** [comment]
Yes, there are two key bits of information that are read out of the map header which are missing in your example.

1) is the location of the observer.
2) the radius of the sun as assumed by AIA.

What is actually happening under the hood is slightly more subtle than having missing information though. You are asking Map to convert a world coordinate in a HPC frame which is not the one described by the Map (it assumes Earth) to the pixel frame of the map, so what is happening is:

Input SkyCoord -> Map SkyCoord -> Map Pixels

because the observer is not the same between the input frame and the Map frame, to do the first conversion it is doing

Input HPC -> HGS -> Map HPC to correctly account for the change in origin of the two HPC frames. This transformation is only valid for points on the disk (as you can't calculate distance off disk) which is why you end up with NaN.

The answer to you issue is you should always use `map.coordinate_frame` when wanting to use `world_to_pixel` as it is the native frame of the Map

**Comment by WaaallEEE** [comment]
Thank you @Cadair for this 1st clarification. Indeed, I was seeing the difference in observer when doing frame = frames.Helioprojective was associated with having NaNs but didn't understand why it is necessary to have an intermediate conversion to HGS which makes the whole conversion trip for off limb pixels. 
I think we misunderstood how Sunpy defines the "observer".  Here's the difference between using `frames.Helioprojective` and `aia_map.frame` to define the coordinates. 

```
xy_coord1
Out[22]: 
<SkyCoord (Helioprojective: obstime=2011-06-07 06:33:02.770000, rsun=696000000.0 m, observer=<HeliographicStonyhurst Coordinate (obstime=2011-06-07 06:33:02.770000): (lon, lat, radius) in (deg, deg, m)
    (0., 0.048591, 1.51846026e+11)>): (Tx, Ty) in arcsec
    (-167., 931.)>
xy_coord2
Out[23]: 
<SkyCoord (Helioprojective: obstime=2011-06-07 06:33:02.770000, rsun=695508.0 km, observer=<HeliographicStonyhurst Coordinate (obstime=2011-06-07 06:33:02.770000): (lon, lat, radius) in (deg, deg, AU)
    (0., 0.03652003, 1.01484379)>): (Tx, Ty) in arcsec
    (-167., 931.)>
```

So is the observer assumed to be Earth when we use `obstime` with `frame=frames.Helioprojective`?
If yes, in the original code above does the system trip because we use the `world_to_pixel()` from an AIA map object and not from an earth-based map object while giving an earth-based frame?

**Comment by Cadair** [comment]
> So is the observer assumed to be Earth when we use `obstime` with `frame=frames.Helioprojective`?

Correct, the observer defaults to `'earth'` and specifying obstime allows us to calculate the position of the Earth.

> If yes, in the original code above does the system trip because we use the `world_to_pixel()` from an AIA map object and not from an earth-based map object while giving an earth-based frame?

Also correct :smile:. The input coordinate (in any frame) will be converted to the native frame of the Map before doing the pixel transformation.

**Comment by WaaallEEE** [comment]
Thank you again @Cadair . 
Last one (a very easy one): for efficiency we must avoid reading in the pixels in fits or jpeg2000 file to build a map, we just want to use the headers to benefit from the whole conversion framework. If we use a dummy array (say `np.zeros([4096, 4096]`), and wcs object when initializing the sunpy map, is the dummy array hard-copied in the map or will it be a numpy view of it? 

Either way, do you see a smarter way?

**Comment by Cadair** [comment]
It's a much better idea to initialise the Map with the header directly and not the WCS object, because a WCS object dosen't always have all the infomation needed that's in the header.

You can use `np.zeros` although `np.empty` would be faster. Also, you don't need to make the array the same size as it's supposed to be if you just want to use `world_to_pixel` and stuff. so `sunpy.map.Map(np.empty(()), aia_header)` should work fine. Alternatively you can memmap the the data from the FITS file, so that it's not all read into memory.

**Comment by WaaallEEE** [comment]
@Cadair 
Thanks but question not quite answered
> initializing the sunpy map, is the dummy array hard-copied in the map or will it be a numpy view of it?
(assuming it's not an np.empty array)

If each time it's just a view that's used, and not a hard copy, then I can as well define one array (empty or not) once and for all (for thousands of files) instead of redefining an np.empty each time. Hence that question.

**Comment by Cadair** [comment]
It wont copy if you don't run it through anything like `rotate` or similar. The `data` object dosen't actually need to even be a numpy array. I have successfully used other array types like Dask. Having said that I don't know what Map expects the array to provide for `world_to_pixel`

**Comment by ayshih** [comment]
If I understand correctly, the initial issue was that @WaaallEEE did not understand that this line:
```
xy_coord2 = SkyCoord(*xy_hpc, obstime=aia_map.date, frame=frames.Helioprojective)
```
assumes an Earth-based observer, and that instead he needed to specify the correct observer by using:
```
xy_coord2 = SkyCoord(*xy_hpc, frame=frames.Helioprojective(obstime=aia_map.date, observer=aia_map.observer_coordinate)
```
or, since he wanted the exact same frame as the original map, simply:
```
xy_coord2 = SkyCoord(*xy_hpc, frame=aia_map.frame)
```

Are we going to add anything to documentation, or should we just close this issue?

**Comment by ayshih** [comment]
A consequence of the now-merged #3388 is that users will no longer get thrown off by `Helioprojective` having a default observer of Earth.  The call `aia_map.world_to_pixel(xy_coord2)` in the original post will now raise a `ConvertError` because `xy_coord2` doesn't have a specified observer – and now no default observer – and hence can't be converted to the coordinate frame of `aia_map`.

Closing this issue
